docs/proposals/26-workload-identity.md

```diff
@@ -0,0 +1,514 @@
+feature that Gardener can benefit of as well because it will not have to store
+the credentials of Gardener users.
+
+### Goals
+
+- Manage shoot clusters without credentials provided by the user.
+- Replace static credentials for gardener system components, e.g. DNS and backup
+  controllers.
+- Rotate credentials regularly. Rotating the token signing keys, will
+  effectively invalidate all previously issued tokens.
+- Offload Gardener users from the burden to store and manage static credentials
+  for their accounts.
+
+### Non-Goals
+
+- Register gardener as trusted identity in the shoot clusters.
+- The tokens to be usable for authentication with the Gardener API.
+- Compatibility with gardenctl integration with cloud provider CLIs.
+- OIDC compliance - just as Kubernetes, Gardener goal is not to have full OIDC
+  compliance, but to implement the bare minimum for OIDC compatible trust
+  federation.
+
+## Proposal
+
+In short, Gardener API server will generate JWTs on request by gardenlet.
+Gardenlet will ensure that the tokens can be consumed conveniently by the
+various components in the seed clusters, for example will write them in secrets.
+Service provider extensions will be responsible for making adjustments so that
+the token is consumable by the service SDK. For shoot clusters, it would be
+natural that the `cloudprovider` secret is reused as storage target for the
+token.
+
+Gardenlet will take care to refresh the token regularly and on time so that the
+target storage always contains a valid token. As an example a token can be
+refreshed when it reaches 80% of its lifetime. Of course it will all depend on
+the validity duration of such tokens. This is why these parameters will be
+configurable.
+
+The OIDC metadata discovery documents will be served in such network segment,
+e.g. public internet, so that service providers can be configured to trust
+Gardener as an OIDC compatible token issuer.
+
+### API Changes
+
+A new resource `WorkloadIdentity` in `security.gardener.cloud` API Group will be
+implemented. It will specify different characteristics of the JWT, like the
+value for the `aud` claim. The `WorkloadIdentity` resource will allow the token
+duration to be set by Gardener users so they can use tokens with shorter or
+longer validity compared to the default one. This duration will be ensured to be
+between certain limits of minimal and maximal validity, in order to avoid
+frequent token renewals as well as tokens with too long validity.
+
+Similarly to `providerConfig` in other APIs, `WorkloadIdentity` resource will
+feature a `config` field that will be of byte array type allowing service
+provider specific configurations. Usually, the clients for services supporting
+identity federation needs additional information about the service account and
+the federated identity in order to successfully use the JWT. This information is
+known to the service account owners and they will provide it via this `config`
```

Comment:
Then let's call it `providerConfig` to be in line with other APIs.

Reply:
Done with d2eda46ee9d1c0